src/test/queue.spec.ts

```diff
@@ -0,0 +1,146 @@
+import * as sinon from "sinon";
+import * as chai from "chai";
+import {fail} from 'assert';
+
+chai.use(require("chai-as-promised"));
+const { expect } = chai;
+
+const Queue = require("../queue");
+
+const TEST_ERROR = new Error("foobar");
+
+describe("Queue", () => {
+  it("should ignore non-number backoff", () => {
+    const q = new Queue({
+      backoff: "not a number"
+    });
+    expect(q.backoff).to.equal(200);
+  });
+
+  it("taskName should be itself for string task", () => {
+    const handler = sinon.stub().resolves();
+    const q = new Queue({
+      handler,
+    });
+
+    const stringTask = "test task";
+    q.add(stringTask);
+
+    expect(q.taskName(0)).to.equal(stringTask);
+  });
+
+  it("taskName should be an index for non-string task", () => {
+    const handler = sinon.stub().resolves();
+    const q = new Queue({
+      handler,
+    });
+
+    const nonStringTask = 2;
+    q.add(nonStringTask);
+
+    expect(q.taskName(0)).to.equal("index 0");
+  });
+
+  it("should handle tasks", () => {
+    const handler = sinon.stub().resolves();
+    const q = new Queue({
+      handler,
+    });
+
+    q.add(4);
+    q.close();
+
+    return q.wait()
+      .then(() => {
+        expect(handler.callCount).to.equal(1);
+        expect(q.complete).to.equal(1);
+        expect(q.success).to.equal(1);
+        expect(q.errored).to.equal(0);
+        expect(q.retried).to.equal(0);
+      });
+  });
+
+  it("should not retry", () => {
+    const handler = sinon.stub().rejects(TEST_ERROR);
+    const q = new Queue({
+      handler,
+      retries: 0,
+    });
+
+    q.add(4);
+    q.close();
+
+    return q.wait()
+      .then(() => {
+        throw new Error("handler should have rejected");
+      })
+      .catch((err: Error) => {
+        expect(err).to.equal(TEST_ERROR);
+      })
+      .then(() => {
+        expect(handler.callCount).to.equal(1);
+        expect(q.complete).to.equal(1);
+        expect(q.success).to.equal(0);
+        expect(q.errored).to.equal(1);
+        expect(q.retried).to.equal(0);
+      });
+  });
+
+  it("should retry the number of retries, plus one", () => {
+    const handler = sinon.stub().rejects(TEST_ERROR);
+    const q = new Queue({
+      backoff: 0,
+      handler,
+      retries: 3,
+    });
+
+    q.add(4);
+    q.close();
+
+    return q.wait()
+      .then(() => {
+        throw new Error("handler should have rejected");
+      })
+      .catch((err: Error) => {
+        expect(err).to.equal(TEST_ERROR);
+      })
+      .then(() => {
+        expect(handler.callCount).to.equal(4);
+        expect(q.complete).to.equal(1);
+        expect(q.success).to.equal(0);
+        expect(q.errored).to.equal(1);
+        expect(q.retried).to.equal(3);
+      });
+  });
+
+  it("should retry the number of retries for both tasks", () => {
+    const stub = sinon.stub()
```

Comment:
The way this is written I think will cause indeterministic behavior in the tests. I think this should be adjusted this way:

```javascript
const helper = sinon.stub()
  .rejects(TEST_ERROR) //  by default, rejects
  .withArgs("one").onCall(2).resolves() // when called with "one", resolves the second time
  .withArgs("two").onCall(3).resolves(); // when called with "two", resolves the third time
```

Also:
- remove `const handler = stub;` - just rename `stub` to `handler`
- `q.add` now needs to be provided `"one"` and `"two"`, which will trigger these specified behaviors.

This _probably_ can help solve the question of how to test concurrency too, since you can have more fine grained control over the behavior.

That all make sense?

Reply:
For this test, I think there should not be any flakiness as long as the concurrency is set to 1.

I can use the `withArgs` to create some tests for concurrency cases.